### DISCUSSION TITLE
Cannot Follow Users

### DISCUSSION THREAD

**Discussion by badideasforsale** [mastodon/mastodon#22356] [Q&A]
### Steps to reproduce the problem

I set up a new instance
1. Set up a brand-new Mastodon instance ([hosted on AWS](https://github.com/widdix/mastodon-on-aws))
2. Created a user and logged in
3. Used `tootctl` to confirm email

Then began to follow some users:
1. Started to follow some accounts
1. Nothing was appearing in my feed
1. Profile showed I was not following anyone
1. Checked logs

### Expected behaviour

Followed someone

### Actual behaviour

Did not follow someone

### Detailed description

I see a number of failed jobs in Sidekiq
<img width="987" alt="Screenshot 2022-12-13 at 7 13 27 PM" src="https://user-images.githubusercontent.com/80412585/207496236-f72b444b-ba85-4252-bd24-5635feb13562.png">
<img width="271" alt="image" src="https://user-images.githubusercontent.com/80412585/207496461-fcf74d13-5ac3-457c-a025-3b598885a327.png">

I see in the logs a ton of errors, all appearing to have a similar message: `"error_message": "Couldn't find Account with 'id'=109509462473004442"` where the number corresponds to someone I was attempting to follow. This happens with everyone I try to follow, regardless of their home instance.

Log excerpt:

```
2022-12-14T02:09:02.051Z pid=6 tid=2hcu WARN: {
    "context": "Job raised exception",
    "job": {
        "retry": 16,
        "queue": "push",
        "dead": false,
        "args": [
            "{\"@context\":\"https://www.w3.org/ns/activitystreams\",\"id\":\"https://social.securitytheater.net/68f3251c-05b0-4bf1-9d01-3f099f81a16a\",\"type\":\"Follow\",\"actor\":\"https://social.securitytheater.net/users/spaceinvader\",\"object\":\"https://fiasco.social/users/dbsmasher\"}",
            109509462473004442,
            "https://fiasco.social/users/dbsmasher/inbox"
        ],
        "class": "ActivityPub::DeliveryWorker",
        "jid": "c73805f10d447825cb09ba4f",
        "created_at": 1670983742.0455234,
        "enqueued_at": 1670983742.045663
    }
}

2022-12-14T02:09:52.697Z pid=6 tid=2h9i WARN: {
    "context": "Job raised exception",
    "job": {
        "retry": 16,
        "queue": "push",
        "dead": false,
        "args": [
            "{\"@context\":\"https://www.w3.org/ns/activitystreams\",\"id\":\"https://social.securitytheater.net/68f3251c-05b0-4bf1-9d01-3f099f81a16a\",\"type\":\"Follow\",\"actor\":\"https://social.securitytheater.net/users/spaceinvader\",\"object\":\"https://fiasco.social/users/dbsmasher\"}",
            109509462473004442,
            "https://fiasco.social/users/dbsmasher/inbox"
        ],
        "class": "ActivityPub::DeliveryWorker",
        "jid": "c73805f10d447825cb09ba4f",
        "created_at": 1670983742.0455234,
        "enqueued_at": 1670983792.6916714,
        "error_message": "Couldn't find Account with 'id'=109509462473004442",
        "error_class": "ActiveRecord::RecordNotFound",
        "failed_at": 1670983742.0498407,
        "retry_count": 1,
        "retried_at": 1670983761.4005384
    }
}
```

### Specifications

```
Mastodon	4.0.2
Ruby 		 3.0.4p208
PostgreSQL	14.5
Redis 		5.0.6
```

**Comment by ClearlyClaire** [comment]
In these errors, 109509462473004442 would be your own `id`. So it seems like either you deleted a previous local account you have tried to use and this is a red herring, or `mastodon-web` and `mastodon-sidekiq` are not sharing the same PostgreSQL database. Do you think that may be it?

**Comment by badideasforsale** [comment]
I believe `mastodon-web` and `mastodon-sidekiq` are both using the same database.  I don't see any database connection errors in any of the logs. Other Sidekiq jobs are running successfully, and the web interface _appears_ to be working correctly (including logins, etc).

When approving my account, I hit a snag, and first tried to approve before realizing I needed to confirm my account. I also first attempted to use the display name, not the username. Here is my shell history:

```
# RAILS_ENV=production bin/tootctl accounts list
Could not find command "list".
# RAILS_ENV=production bin/tootctl accounts
   << help text snipped >>
# RAILS_ENV=production bin/tootctl accounts approve 👾👾👾
No such account
# RAILS_ENV=production bin/tootctl accounts approve help
No such account
# RAILS_ENV=production bin/tootctl accounts help
   << help text snipped >>
# RAILS_ENV=production bin/tootctl accounts approve spaceinvader
OK
# RAILS_ENV=production bin/tootctl accounts approve spaceinvader
OK
# RAILS_ENV=production bin/tootctl accounts modify spaceinvader --confirm
OK
# RAILS_ENV=production bin/tootctl accounts modify spaceinvader --role Owner
OK
```

Later I ran `# RAILS_ENV=production bin/tootctl settings registrations approved --require-reason`.

I have never deleted an account on this instance. I don't actually know how to (yet) although I reckon by RTFM-ing I'll figure it out.

Should I use `tootctl` to delete my local `spaceinvader` account and set up a new one? It doesn't look like changing usernames is possible at the moment. In the meantime, since this seems to be a local-server-issue, I'll also try setting up a different user account to ensure the overall system is healthy and rule out a bad setup.

**Comment by badideasforsale** [comment]
## What I did now
I tried to create a new user, `toor`. I signed up via the web form, but because this host isn't capable of sending email (yet) I logged in and approved by new user via `tootctl`. Same issue as before. I've got some log lines here that are hopefully useful.

## Email
Here's the message being enqueued:
```
[ActiveJob] [ActionMailer::MailDeliveryJob] [2589de5b-9bb1-4dc3-95e5-0e1b6f25be2a] Performing ActionMailer::MailDeliveryJob (Job ID: 2589de5b-9bb1-4dc3-95e5-0e1b6f25be2a) from Sidekiq(mailers) enqueued at 2022-12-16T02:45:40Z with arguments: "UserMailer", "confirmation_instructions", "deliver_now", {:args=>[#<GlobalID:0x00007f4bd6411bc0 @uri=#<URI::GID gid://mastodon/User/2>>, "QpXrdbTyCBxL6CcZiyWc"]}
```
...and here's Sidekiq unhappy. I don't know if this is expected unhappiness due to broken email or not.
```
[ActiveJob] [ActionMailer::MailDeliveryJob] [2589de5b-9bb1-4dc3-95e5-0e1b6f25be2a] Error performing ActionMailer::MailDeliveryJob (Job ID: 2589de5b-9bb1-4dc3-95e5-0e1b6f25be2a) from Sidekiq(mailers) in 3.56ms: NoMethodError (undefined method `memorial?' for nil:NilClass):
```

## User approval
Here I'm approving, confirming, and setting this new user up as Owner of my Mastodon instance.
```
# RAILS_ENV=production bin/tootctl accounts approve toor
OK
# RAILS_ENV=production bin/tootctl accounts modify toor --confirm
OK
# RAILS_ENV=production bin/tootctl accounts modify toor --role Owner
OK
```

## Cannot follow other users
I imagine this is my new user `id`, also failing in the same manner:
```
2022-12-16T02:46:46.349Z pid=6 tid=2qoq WARN: ActiveRecord::RecordNotFound: Couldn't find Account with 'id'=109521059179025980
```

In the GUI, I'm logged in, can browse around, edit preferences, etc.  That at least appears to be working correctly, and communicating with the database. (I see little wiggles in load when logging in, too.)  Plus, Sidekiq has been busy notching over 8000 successful jobs... so I think that means it is communicating to the database.

**Reply by ClearlyClaire** [reply]
What happens if you do `Account.find(109521059179025980)` in a Rails console (`RAILS_ENV=production bundle exec rails c`)?

I'm not familiar with AWS nor the specific config you are using, but I still suspect sidekiq and puma are somehow using different databases.

**Reply by badideasforsale** [reply]
I see the user 🤷‍♂️ 
```
irb(main):001:0> Account.find(109521059179025980)
=>
#<Account:0x00007f5230c63ea8
 username: "toor",
 domain: nil,
 private_key:
  "<snip>",
 public_key:
  "<snip>",
 created_at: Fri, 16 Dec 2022 02:45:39.928722000 UTC +00:00,
 updated_at: Fri, 16 Dec 2022 02:45:39.928722000 UTC +00:00,
 note: "",
 display_name: "toor",
 uri: "",
 url: nil,
 avatar_file_name: nil,
 avatar_content_type: nil,
 avatar_file_size: nil,
 avatar_updated_at: nil,
 header_file_name: nil,
 header_content_type: nil,
 header_file_size: nil,
 header_updated_at: nil,
 avatar_remote_url: nil,
 locked: false,
 header_remote_url: "",
 last_webfingered_at: nil,
 inbox_url: "",
 outbox_url: "",
 shared_inbox_url: "",
 followers_url: "",
 protocol: "ostatus",
 id: 109521059179025980,
 memorial: false,
 moved_to_account_id: nil,
 featured_collection_url: nil,
 fields: nil,
 actor_type: nil,
 discoverable: nil,
 also_known_as: nil,
 silenced_at: nil,
 suspended_at: nil,
 hide_collections: nil,
 avatar_storage_schema_version: nil,
 header_storage_schema_version: nil,
 devices_url: nil,
 sensitized_at: nil,
 suspension_origin: nil,
 trendable: nil,
 reviewed_at: nil,
 requested_review_at: nil>
```

**Reply by ClearlyClaire** [reply]
Are you running that from the machine/VM running sidekiq, or the one running puma (mastodon-web), or yet another one? It seems to suggest sidekiq is not using the same database, as if they were both provisioned with their own PostgreSQL database (so you wouldn't have PostgreSQL connection errors, but the database would not have the same data at all, and jobs queued from mastodon-web would refer to objects sidekiq does not know about).
